### PR TITLE
Move freeplay links to the right and extract header.js.erb

### PR DESCRIPTION
**Pull request by joshlory**
Move free play links next to the sign-in button per @sfilman's mock:

screen shot 2015-01-27 at 1 24 56 pm

This required extracting JS from `_header_progress.html.haml` into its own file.

### CODE REVIEW

File: dashboard/app/assets/javascripts/header.js.erb [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+/**
+ * @param scriptData{{
+ *   finishLink: string,
+ *   statsPath: string,
+ *   showStageLinks: boolean,
+ *   trophies: Object,
+ *   currentLevelIndex: number,
+ *   scriptId: number,
+ *   scriptLevelId: number,
+ *   levels: Array.<{
```
Comment (joshlory) [inline dashboard/app/assets/javascripts/header.js.erb]: GitHub highlights this on the diff but it's inside a comment and correct per https://developers.google.com/closure/compiler/docs/js-for-compiler#types

File: dashboard/app/views/shared/_user_header.html.haml [hunk 2]
```diff
@@ -1,5 +1,4 @@
 - if current_user
-  .transparent_backdrop{style: 'display:none;'}
```
Comment (laurelfan) [inline dashboard/app/views/shared/_user_header.html.haml]: do you know what this was for?

File: dashboard/app/views/layouts/_header.html.haml [hunk 3]
```diff
@@ -25,6 +25,18 @@
               = render partial: 'shared/header_progress'
             - elsif @edit_blocks
               .header_level{style:'font-size: 250%; line-height: 125%;'} Editing #{@edit_blocks.humanize}
-          - if params[:controller] !~ /^sessions/
-            .header_user
-              = render partial: 'shared/user_header'
+          .header_right
+            %span.stage_links{style: 'display: none;'}
```
Comment (laurelfan) [inline dashboard/app/views/layouts/_header.html.haml]: not sure if stage_links sounds like the right name for this... standalone_level_links?

IIRC stage_links was used to share CSS with something else but it doesn't seem to be used anymore.

Comment (laurelfan) [inline dashboard/app/views/layouts/_header.html.haml]: for that matter is the whole logic for when we show this still correct? Do we want to show these links all the time now?

File: dashboard/app/views/shared/_user_header.html.haml (hunk 2 above)
Comment (joshlory) [inline dashboard/app/views/shared/_user_header.html.haml]: Removed in https://github.com/code-dot-org/code-dot-org/pull/1087 but I missed this one.

File: dashboard/app/views/layouts/_header.html.haml (hunk 3 above)
Comment (joshlory) [inline dashboard/app/views/layouts/_header.html.haml]: They're `display: none;` by default, then they get shown based on the `scriptData` JSON here: https://github.com/code-dot-org/code-dot-org/pull/1154/files#diff-3bcf40f621bb1e4fc4cb234be362c2a9L40

Comment (joshlory) [inline dashboard/app/views/layouts/_header.html.haml]: Renamed to `.freeplay_links` in 14e9c75

Comment (laurelfan) [inline dashboard/app/views/layouts/_header.html.haml]: Hm, I think in Sarah's mock, it's ambiguous where the freeplay_links should show up. I see them on slide 2 which is the homepage (which is why I was thinking that the mocks were showing them all the time) but do not see them on slide 7 which is also the homepage. Maybe ask her?

Comment (joshlory) [inline dashboard/app/views/layouts/_header.html.haml]: This change doesn't change when the free play links show up.  It looks like it's defined as `script.twenty_hour? || script.stages.to_a.count > 1`.

Comment (laurelfan) [inline dashboard/app/views/layouts/_header.html.haml]: Right, looks like Sarah's mock intends to change where they show up though...

**Comment by sfilman** [comment]
slide 7 was a mock bug -- intent was that on all studio pages, the header has the buttons and they are moved to the right next to the identity button.

**Comment by breville** [comment]
Please make sure this works on all browser widths, even down to portrait
mobile devices, etc.

**Comment by joshlory** [comment]
To reiterate, this change only moves the existing icons to a different part of the header.  It doesn't add or remove icons, or change which pages the icons appear on.  @breville is there a specific device you have concerns about?  We don't currently show the header on devices smaller than 900x500.

**Comment by breville** [comment]
No explicit worries, but the header is used on both responsive (e.g.
tutorial) and non-responsive pages (e.g. studio.code.org) and even on the
latter, the sign-in button moves responsively.  Just do the usual resize
test to eyeball its behaviour still works at a range of resolutions.

That said, doesn't it show these icons on studio.code.org now, whereas it
didn't before?

**Comment by joshlory** [comment]
The studio.code.org homepage header is unchanged:
screen shot 2015-01-28 at 3 02 01 pm

**Pull request merged**